source/sso_token_utils.c

```diff
@@ -0,0 +1,165 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/auth/private/sso_token_utils.h>
+#include <aws/cal/hash.h>
+#include <aws/common/encoding.h>
+#include <aws/common/file.h>
+#include <aws/common/json.h>
+
+#if defined(_MSC_VER)
+#    pragma warning(disable : 4232)
+#endif /* _MSC_VER */
+
+AWS_STATIC_STRING_FROM_LITERAL(s_sso_cache_directory, "/.aws/sso/cache/");
+
+struct aws_string *aws_construct_token_path(struct aws_allocator *allocator, const struct aws_string *input) {
+    AWS_PRECONDITION(input);
+
+    struct aws_string *token_path_str = NULL;
+
+    struct aws_string *home_directory = aws_get_home_directory(allocator);
+    if (!home_directory) {
+        return NULL;
+    }
+
+    struct aws_byte_cursor home_dir_cursor = aws_byte_cursor_from_string(home_directory);
+    struct aws_byte_cursor cache_dir_cursor = aws_byte_cursor_from_string(s_sso_cache_directory);
+    struct aws_byte_cursor input_cursor = aws_byte_cursor_from_string(input);
+    struct aws_byte_cursor json_cursor = aws_byte_cursor_from_c_str(".json");
+
+    struct aws_byte_buf token_path_buf;
+    AWS_ZERO_STRUCT(token_path_buf);
+    struct aws_byte_buf sha1_buf;
+    AWS_ZERO_STRUCT(sha1_buf);
+
+    /* append home directory */
+    if (aws_byte_buf_init_copy_from_cursor(&token_path_buf, allocator, home_dir_cursor)) {
+        goto cleanup;
+    }
+
+    /* append sso cache directory */
+    if (aws_byte_buf_append_dynamic(&token_path_buf, &cache_dir_cursor)) {
+        goto cleanup;
+    }
+
+    /* append hex encoded sha1 of input */
+    if (aws_byte_buf_init(&sha1_buf, allocator, AWS_SHA1_LEN) ||
+        aws_sha1_compute(allocator, &input_cursor, &sha1_buf, 0)) {
+        goto cleanup;
+    }
+    struct aws_byte_cursor sha1_cursor = aws_byte_cursor_from_buf(&sha1_buf);
+    if (aws_hex_encode_append_dynamic(&sha1_cursor, &token_path_buf)) {
+        goto cleanup;
+    }
+
+    /* append .json */
+    if (aws_byte_buf_append_dynamic(&token_path_buf, &json_cursor)) {
+        goto cleanup;
+    }
+
+    // Use platform-specific directory separator.
+    const char local_platform_separator = aws_get_platform_directory_separator();
+    for (size_t i = 0; i < token_path_buf.len; ++i) {
+        if (aws_is_any_directory_separator((char)token_path_buf.buffer[i])) {
+            ((char *)token_path_buf.buffer)[i] = local_platform_separator;
+        }
+    }
```

Comment:
I created the [PR](https://github.com/awslabs/aws-c-common/pull/1010) to normalize paths, but it crashes if someone passes non-writable memory to the function. We can always allocate memory for normalized path but that is not that efficient. I realized that we can just construct a cross-platform path instead of constructing then normalizing.